lib/puppet/property.rb

```diff
@@ -402,6 +402,7 @@ def property_matches?(current, desired)
   # @param value [Object] the value to format as a string
   # @return [String] a pretty printing string
   def is_to_s(value)
+    value = value[0] if value.is_a?(Array) && value.size == 1
```

Comment:
Thomas and I talked this over, but for posterity:

IIRC, one reason for the above idiom was because ruby 1.8.7 used to automatically coerce single element arrays into strings. So that idiom handled that, in addition to the fact that `Puppet::Property` supports both single and multi-valued values, but `@should` is always an array.

About unified `[]` notation, we used to always emit single value arrays in 4.10.x, note `['wheel']` in the output below:

```
[root@xqf43z9asakgxjg ~]# puppet --version
4.10.1
[root@xqf43z9asakgxjg ~]# puppet apply -e "user { 'foobar': ensure => present }"
Notice: Compiled catalog for xqf43z9asakgxjg.delivery.puppetlabs.net in environment production in 0.08 seconds
Notice: /User[foobar]/ensure: created
Notice: Applied catalog in 0.06 seconds
[root@xqf43z9asakgxjg ~]# puppet apply -e "user { 'foobar': ensure => present, groups => ['wheel'] }"
Notice: Compiled catalog for xqf43z9asakgxjg.delivery.puppetlabs.net in environment production in 0.08 seconds
Notice: /Stage[main]/Main/User[foobar]/groups: groups changed '' to ['wheel']
Notice: Applied catalog in 0.06 seconds
```

But we are not consistent about array-bracketing the `current` and `desired` values. In the example, note `groups changed 'wheel,sshd' to ['mail', 'sshd', 'wheel']`. The `current` value should be bracketed too.

```
[root@xqf43z9asakgxjg ~]# puppet apply -e "user { 'foobar': ensure => present, groups => ['wheel', 'sshd'] }"
Notice: Compiled catalog for xqf43z9asakgxjg.delivery.puppetlabs.net in environment production in 0.07 seconds
Notice: /Stage[main]/Main/User[foobar]/groups: groups changed 'wheel' to ['sshd', 'wheel']
Notice: Applied catalog in 0.05 seconds
[root@xqf43z9asakgxjg ~]# puppet apply -e "user { 'foobar': ensure => present, groups => ['wheel', 'sshd', 'mail'] }"
Notice: Compiled catalog for xqf43z9asakgxjg.delivery.puppetlabs.net in environment production in 0.08 seconds
Notice: /Stage[main]/Main/User[foobar]/groups: groups changed 'wheel,sshd' to ['mail', 'sshd', 'wheel']
Notice: Applied catalog in 0.06 seconds
```

Part of the problem is that the `user` type's `groups` property relies on `Puppet::Property::List` which overrides the default `is_to_s` logic to use different delimiters.